### ISSUE TITLE
Crash on lambda parameter

### ISSUE THREAD

**Issue by CypherGrue** [Technologicat/pyan#61]
Running pyan 1.1.1 on following trivial example
` def func(x=lambda a,b:a):
    return x(1,2)
`

Gives following error
>   File "/usr/local/lib/python3.9/site-packages/pyan/analyzer.py", line 468, in visit_Lambda
    with ExecuteInInnerScope(self, "lambda"):
  File "/usr/local/lib/python3.9/site-packages/pyan/anutils.py", line 224, in __enter__
    raise ValueError("Unknown scope '%s'" % (inner_ns))
ValueError: Unknown scope 'test2.func.lambda'

**Comment by Technologicat** [comment]
Hi,

It seems I've missed a corner case, and it's analyzing the `lambda` in the wrong scope. Good catch!

The error manifests in this particular way, because the scope analysis (see `analyze_scopes`), which uses the `symtable` stdlib module, populates the scope hierarchy correctly, but the analyzer proper makes a mistake when it thinks the lambda is inside the function scope.

The actual problem is that in `visit_FunctionDef`, the call to `analyze_arguments` occurs after the analyzer enters the function scope.

Fixing this is a bit complicated due to how the analyzer is implemented. The call to `generate_args_nodes` must run inside the function scope, so that the names being defined as function parameters will be scoped correctly. However, the *values* should be analyzed in the context of the surrounding scope. Currently there's no easy way to make the analyzer do this.

There's also a related old (but still live) bug that Pyan doesn't currently distinguish between different lambdas in the same parent scope, because everything needs a name for the call graph, and all lambdas are named "lambda". Right now I don't recall if `symtable` names the scopes like that, or if Pyan itself does. If the latter, we should be able to use the `lineno` attribute from the AST node to uniqify the names. But in the worst case, we need to reimplement what `symtable` is already doing. Scope analysis in Python is [rather painful](https://cs.brown.edu/~sk/Publications/Papers/Published/pmmwplck-python-full-monty/paper.pdf).

So, in summary: thanks for reporting! Fixing this requires a bit more work than I have time for right now, but this is definitely something to fix in a future release.

**Comment by akashkvyas** [comment]
Hello,

I've been trying to understand how the analyzer is scoping the values for lambda's in arguments, like the example above. My approach was to write a recursive decorator that wraps the ast visitors, preventing the analyzer from entering the function scope, but still allowing it to call analyze_arguments. I'm not sure how well this would work, as I have been writing some tests but am running into a lot of issues with the custom recursive implementation. Here's a rough idea of what I was thinking, please let me know what you think!
```
def recursive(func):
    """Recursive wrapper to allow evaluation in different scopes"""
    def wrapper(self, node):
        if node.name == "function":
            # analyze arguements before entering function scope
            analyze_arguments(nodes.args)
        for child in ast.iter_child_nodes(node):
            self.visit(child)
    return wrapper
```

**Comment by nguerette** [comment]
As a note of additional info: the same problem seems to exist for list comprehensions.
